File: calendar-server/src/test/java/uk/co/unclealex/hammers/calendar/server/dates/DateServiceImplTest.java
/**
 * Copyright 2010-2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 */

package uk.co.unclealex.hammers.calendar.server.dates;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import java.util.Map;
import java.util.Map.Entry;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.junit.Test;

import com.google.common.base.Strings;
import com.google.common.collect.Maps;

/**
 * The Class DateServiceImplTest.
 * 
 * @author alex
 */
public class DateServiceImplTest {

  /**
   * Test next year.
   */
  @Test
  public void testNextYear() {
    testYear(dayOf(6, 1), dateOf(25, 12, 2012), false, dateOf(6, 1, 2013));
  }

  /**
   * Test year at expecting later.
   */
  @Test
  public void testThisYearAtExpectingLater() {
    testYear(dayOf(6, 1), dateOf(5, 1, 2013), false, dateOf(6, 1, 2013));
  }

  /**
   * Test year at expecting earlier.
   */
  @Test
  public void testThisYearAtExpectingEarlier() {
    testYear(dayOf(6, 1), dateOf(5, 9, 2012), true, dateOf(6, 1, 2012));
  }

  /**
   * Test last year.
   */
  @Test
  public void testLastYear() {
    testYear(dayOf(25, 12), dateOf(6, 1, 2012), true, dateOf(25, 12, 2011));
  }

  /**
   * Test year aware parse.
   */
  @Test
  public void testYearAwareParse() {
    testParseAndFind(minuteOf(5, 9, 1972, 9, 12), "05/09/1972 9:12", dateOf(10, 10, 2012), true, "dd/MM[/yyyy] HH:mm");
  }

  /**
   * Test possibly yearless parse.
   */
  @Test
  public void testPossiblyYearlessParse() {
    testParseAndFind(minuteOf(5, 9, 2012, 9, 12), "05/09 9:12", dateOf(10, 10, 2012), true, "dd/MM[/yyyy] HH:mm");
  }

  /**
   * Test definitely yearless parse.
   */
  @Test
  public void testDefinitelyYearlessParse() {
    testParseAndFind(minuteOf(5, 9, 2012, 9, 12), "05/09 9:12", dateOf(10, 10, 2012), true, "dd/MM HH:mm");
  }

  /**
   * Test second pass parse.
   */
  @Test
  public void testSecondPassParse() {
    testParseAndFind(
        minuteOf(5, 9, 2012, 9, 12),
        "05/09 9:12",
        dateOf(10, 10, 2012),
        true,
        "HH:mm dd/MM[/yyyy]",
        "dd/MM[/yyyy] HH:mm");
  }

  /**
   * Test yearless with day of week parse.
   */
  @Test
  public void testYearlessWithDayOfWeekParse() {
    testParseAndFind(minuteOf(26, 1, 2012, 9, 0), "9am Thu 26 Jan", dateOf(18, 2, 2012), true, "ha EEE dd MMM");
  }

  /**
   * Test failure to parse.
   */
  @Test
  public void testFailureToParse() {
    testParseAndFind(null, "05:09 9:12", dateOf(10, 10, 2012), true, "dd/MM[/yyyy] HH:mm");
  }

  /**
   * Test parse and find.
   * 
   * @param expectedDateTime
   *          the expected date time
   * @param date
   *          the date
   * @param yearDeterminingDate
   *          the year determining date
   * @param yearDeterminingDateIsLaterThanTheDate
   *          the year determining date is later than the date
   * @param possiblyYearlessDateFormats
   *          the possibly yearless date formats
   */
  public void testParseAndFind(
      final DateTime expectedDateTime,
      final String date,
      final DateTime yearDeterminingDate,
      final boolean yearDeterminingDateIsLaterThanTheDate,
      final String... possiblyYearlessDateFormats) {
    try {
      final DateTime actualDateTime =
          new DateServiceImpl().parsePossiblyYearlessDate(
              date,
              yearDeterminingDate,
              yearDeterminingDateIsLaterThanTheDate,
              possiblyYearlessDateFormats);
      assertEquals("The wrong date was parsed from '" + date + "'.", expectedDateTime, actualDateTime);
    }
    catch (final UnparseableDateException e) {
      assertNull("A date could not be parsed from '" + date + "'.", expectedDateTime);
    }
    for (int paddingLength = 1; paddingLength < 10; paddingLength++) {
      final String padding = Strings.repeat("x", paddingLength);
      for (final String dateContainingString : new String[] {
          date,
          padding + date,
          date + padding,
          padding + date + padding }) {
        try {
          final DateTime actualDateTime =
              new DateServiceImpl().findPossiblyYearlessDate(
                  dateContainingString,
                  yearDeterminingDate,
                  yearDeterminingDateIsLaterThanTheDate,
                  possiblyYearlessDateFormats);
          assertEquals(
              "The wrong date was found within '" + dateContainingString + "'.",
              expectedDateTime,
              actualDateTime);
        }
        catch (final UnparseableDateException e) {
          assertNull("A date could not be found within '" + dateContainingString + "'.", expectedDateTime);
        }
      }
    }
  }

  /**
   * Test week.
   */
  @Test
  public void testWeek() {
    final DateTime tuesday = dateOf(5, 9, 1972);
    final DateTime wednesday = dateOf(6, 9, 1972);
    final DateTime thursday = dateOf(7, 9, 1972);
    final DateTime friday = dateOf(8, 9, 1972);
    final DateTime saturday = dateOf(9, 9, 1972);
    final DateTime sunday = dateOf(10, 9, 1972);
    final DateTime monday = dateOf(11, 9, 1972);

    for (final DateTime dateTime : new DateTime[] { tuesday, wednesday, thursday, friday, monday }) {
      assertTrue(dateTime + " was not identified as a weekday.", DateTimeImplicits.isWeekday(dateTime));
    }

    for (final DateTime dateTime : new DateTime[] { saturday, sunday }) {
      assertFalse(dateTime + " was wrongly identified as a weekday.", DateTimeImplicits.isWeekday(dateTime));
    }
  }

  /**
   * Test3pm saturday.
   */
  @Test
  public void test3pmSaturday() {
    final Map<DateTime, Boolean> expectedResultsByDateTime = Maps.newLinkedHashMap();
    expectedResultsByDateTime.put(minuteOf(9, 9, 1972, 15, 0), true);
    expectedResultsByDateTime.put(minuteOf(9, 9, 1972, 12, 0), false);
    expectedResultsByDateTime.put(minuteOf(9, 9, 1972, 15, 30), false);
    expectedResultsByDateTime.put(minuteOf(9, 9, 1972, 17, 30), false);
    expectedResultsByDateTime.put(minuteOf(10, 9, 1972, 15, 0), false);
    expectedResultsByDateTime.put(minuteOf(10, 9, 1972, 12, 0), false);
    expectedResultsByDateTime.put(minuteOf(10, 9, 1972, 15, 30), false);
    expectedResultsByDateTime.put(minuteOf(10, 9, 1972, 17, 30), false);
    for (final Entry<DateTime, Boolean> entry : expectedResultsByDateTime.entrySet()) {
      final DateTime dateTime = entry.getKey();
      final boolean expectedResult = entry.getValue();
      final boolean actualResult = DateTimeImplicits.isThreeOClockOnASaturday(dateTime);
      assertEquals(
          "The wrong result was found checking whether " + dateTime + " is 3pm on a Saturday.",
          expectedResult,
          actualResult);
    }
  }

  /**
   * Create a new date time.
   * 
   * @param day
   *          The day.
   * @param month
   *          The month.
   * @param year
   *          The year,
   * @param hour
   *          The hour.
   * @param minute
   *          The minute.
   * @return A {@link DateTime} for the supplied information.
   */
  protected DateTime minuteOf(final int day, final int month, final int year, final int hour, final int minute) {
    return new DateTime(year, month, day, hour, minute, 0, 0, DateTimeZone.forID("Europe/London"));
  }

  /**
   * Create a new date time.
   * 
   * @param day
   *          The day.
   * @param month
   *          The month
   * @param year
   *          The year
   * @return A {@link DateTime} for the supplied information.
   */
  protected DateTime dateOf(final int day, final int month, final int year) {
    return minuteOf(day, month, year, 0, 0);
  }

  /**
   * Create a new date time with an unspecified year.
   * 
   * @param day
   *          The day.
   * @param month
   *          The month
   * @return A {@link DateTime} for the supplied information.
   */
  protected DateTime dayOf(final int day, final int month) {
    return new DateTime(1970, month, day, 0, 0, 0, 0, DateTimeZone.forID("Europe/London"));
  }

  /**
   * Test year.
   * 
   * @param dateTime
   *          the date time
   * @param yearDeterminingDate
   *          the year determining date
   * @param yearDeterminingDateIsLaterThanTheDate
   *          the year determining date is later than the date
   * @param expectedDateTime
   *          the expected date time
   */
  protected void testYear(
      final DateTime dateTime,
      final DateTime yearDeterminingDate,
      final boolean yearDeterminingDateIsLaterThanTheDate,
      final DateTime expectedDateTime) {
    final DateTime actualDateTime =
        YearSettingDateParserFactory.setYear(dateTime, yearDeterminingDate, yearDeterminingDateIsLaterThanTheDate);
    assertEquals("The wrong date time was produced.", expectedDateTime, actualDateTime);
  }
}

File: src/test/scala/uk/co/unclealex/hammers/calendar/server/dates/DateServiceImplTest.scala
/**
 * Copyright 2010-2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

package uk.co.unclealex.hammers.calendar.server.dates;

import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import org.specs2.mutable.Specification
import uk.co.unclealex.hammers.calendar.server.January
import uk.co.unclealex.hammers.calendar.server.MonthAndDay
import uk.co.unclealex.hammers.calendar.server.June
import uk.co.unclealex.hammers.calendar.server.March
import uk.co.unclealex.hammers.calendar.server.December
import uk.co.unclealex.hammers.calendar.server.Date
import uk.co.unclealex.hammers.calendar.server.September
import uk.co.unclealex.hammers.calendar.server.February
import uk.co.unclealex.hammers.calendar.server.Instant
import uk.co.unclealex.hammers.calendar.server.October
import com.google.common.base.Strings
import uk.co.unclealex.hammers.calendar.server.dates.DateTimeImplicits._

/**
 * The Class DateServiceImplTest.
 *
 * @author alex
 */
class DateServiceImplTest extends Specification {

  "A day in a month before the current date but meant to be later than it" should {
    "be interpreted as a date in the next year" in {
      January(6) after December(25, 2012) must be equalTo (January(6, 2013))
    }
  }

  "A day in a month after the current date and meant to be later than it" should {
    "be interpreted as a date in the same year" in {
      March(6) after February(25, 2013) must be equalTo (March(6, 2013))
    }
  }

  "A day in a month before the current date and meant to be earlier than it" should {
    "be interpreted as a date in the same year" in {
      June(6) before September(5, 2012) must be equalTo (June(6, 2012))
    }
  }

  "A day in a month after the current date and meant to be earlier than it" should {
    "be interpreted as a date in the previous year" in {
      December(25) before January(6, 2012) must be equalTo (December(25, 2011))
    }
  }

  "A date format with an explicit year" should {
    parseAndFind("05/09/1972 9:12", Before(October(10, 2012)) parsedBy ("dd/MM[/yyyy] HH:mm"),
      Some(September(5, 1972) at (9, 12)))
  }

  "A date format that possibly requires a year to be added" should {
    parseAndFind("05/09 9:12", Before(October(10, 2012)) parsedBy ("dd/MM[/yyyy] HH:mm"),
      Some(September(5, 2012) at (9, 12)))
  }

  "A date format that definitely requires a year to be added" should {
    parseAndFind("05/09 9:12", Before(October(10, 2012)) parsedBy ("dd/MM HH:mm"),
      Some(September(5, 2012) at (9, 12)))
  }

  "A date format that fails the first parse but succedes on the second" should {
    parseAndFind("05/09 9:12", Before(October(10, 2012)) parsedBy ("HH:mm dd/MM[/yyyy]", "dd/MM[/yyyy] HH:mm"),
      Some(September(5, 2012) at (9, 12)))
  }

  "A date format requiring the day of a week but without a year" should {
    parseAndFind("9am Thu 26 Jan", Before(February(18, 2012)) parsedBy ("ha EEE dd MMM"),
      Some(January(26, 2012) at (9, 0)))
  }

  "An invalid date format" should {
    parseAndFind("05:09 9:12", Before(October(10, 2012)) parsedBy ("dd/MM[/yyyy] HH:mm"), None)
  }

  "Checking for whether a day is during the working week or not" should {
    sealed class DayOfWeek(val day: Int, val name: String, val isWeekday: Boolean)
    object Sunday extends DayOfWeek(3, "Sunday", false)
    object Monday extends DayOfWeek(4, "Monday", true)
    object Tuesday extends DayOfWeek(5, "Tuesday", true)
    object Wednesday extends DayOfWeek(6, "Wednesday", true)
    object Thursday extends DayOfWeek(7, "Thursday", true)
    object Friday extends DayOfWeek(8, "Friday", true)
    object Saturday extends DayOfWeek(9, "Saturday", false)
    List(Sunday, Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday) foreach { dayOfWeek =>
      val day = September(dayOfWeek.day, 1972)
      s"${dayOfWeek.name} $day must ${if (dayOfWeek.isWeekday) "" else "not "}be a weekday" in {
        (day.toDateTime isWeekday) must be equalTo (dayOfWeek.isWeekday)
      }
    }
  }

  "Checking for whether an instant is the fabled Saturday at 3pm" should {
    List(
      September(9, 1972) at (15, 0),
      September(9, 1972) at (12, 0),
      September(9, 1972) at (15, 30),
      September(9, 1972) at (19, 45),
      September(10, 1972) at (15, 0),
      September(10, 1972) at (12, 0),
      September(10, 1972) at (15, 30),
      September(10, 1972) at (19, 45)) zip (List(true, false, false, false, false, false, false, false)) foreach {
        case (day, isSaturday3pm) =>
          s"$day must ${if (isSaturday3pm) "" else "not "}be Saturday 3pm" in {
            (day.toDateTime isThreeOClockOnASaturday) must be equalTo (isSaturday3pm)
          }
      }
  }

  /**
   * Test parse and find.
   *
   * @param expectedDateTime
   *          the expected date time
   * @param date
   *          the date
   * @param yearDeterminingDate
   *          the year determining date
   * @param yearDeterminingDateIsLaterThanTheDate
   *          the year determining date is later than the date
   * @param possiblyYearlessDateFormats
   *          the possibly yearless date formats
   */
  def parseAndFind(
    date: String,
    parsingRules: Triple[Boolean, Date, Seq[String]],
    expectedInstant: Option[Instant]) {
    val (yearDeterminingDateIsLaterThanTheDate, yearDeterminingDate, possiblyYearlessDateFormats) = parsingRules
    val expectedDateTime = expectedInstant map (_.toDateTime)
    s"The date string '${date}' using formats ${possiblyYearlessDateFormats.mkString(", ")} must parse to $expectedInstant" in {
      val actualDateTime =
        new DateServiceImpl().parsePossiblyYearlessDate(
          date,
          yearDeterminingDate,
          yearDeterminingDateIsLaterThanTheDate,
          possiblyYearlessDateFormats: _*)
      actualDateTime must be equalTo (expectedDateTime)
    }
    1 to 3 flatMap { paddingSize =>
      val padding = Strings.repeat("x", paddingSize)
      List(date, padding + date, date + padding, padding + date + padding)
    } foreach { date =>
      s"The date string '${date}' using formats ${possiblyYearlessDateFormats.mkString(", ")} must parse to $expectedInstant" in {
        val actualDateTime =
          new DateServiceImpl().findPossiblyYearlessDate(
            date,
            yearDeterminingDate,
            yearDeterminingDateIsLaterThanTheDate,
            possiblyYearlessDateFormats: _*)
        actualDateTime must be equalTo (expectedDateTime)
      }
    }
  }

  /**
   * Syntatic sugar for date parsing information.
   */
  sealed class BeforeOrAfter(date: Date, before: Boolean) {
    def parsedBy(parseStrings: String*) = (before, date, parseStrings)
  }

  case class Before(date: Date) extends BeforeOrAfter(date, true)
  case class After(date: Date) extends BeforeOrAfter(date, false)

  implicit class BeforeAndAfterImplicits(monthAndDay: MonthAndDay) {
    def after(date: Date) = alter(date, false)
    def before(date: Date) = alter(date, true)

    def alter(yearDeterminingDate: Date, yearDeterminingDateIsLaterThanTheDate: Boolean) = {
      val newDateTime = YearSettingDateParserFactory.setYear(
        monthAndDay.toDateTime, yearDeterminingDate.toDateTime, yearDeterminingDateIsLaterThanTheDate)
      Date(newDateTime)
    }

  }
}
